File: Services/PartnerSubscriptionNotificationService.swift
import Foundation
import SwiftUI
import Combine
import FirebaseAuth
import FirebaseFirestore

class PartnerSubscriptionNotificationService: ObservableObject {
    static let shared = PartnerSubscriptionNotificationService()
    
    @Published var shouldShowSubscriptionInheritedMessage = false
    @Published var partnerName = ""
    
    @Published var shouldShowSubscriptionRevokedMessage = false
    @Published var revokedPartnerName = ""
    
    private var cancellables = Set<AnyCancellable>()
    private var partnerListener: ListenerRegistration?
    
    private init() {
        setupObservers()
        startListeningForPartnerSubscription()
    }
    
    private func setupObservers() {
        // Observer les notifications de partage d'abonnement
        NotificationCenter.default.publisher(for: .partnerSubscriptionShared)
            .sink { _ in
                print("🎁 PartnerSubscriptionNotificationService: Notification d'abonnement partagé reçue")
            }
            .store(in: &cancellables)
        
        // Observer les notifications de révocation d'abonnement
        NotificationCenter.default.publisher(for: .partnerSubscriptionRevoked)
            .sink { _ in
                print("🔒 PartnerSubscriptionNotificationService: Notification d'abonnement révoqué reçue")
                // Le partenaire sera notifié via le listener Firebase
            }
            .store(in: &cancellables)
    }
    
    private func startListeningForPartnerSubscription() {
        guard let currentUser = Auth.auth().currentUser else {
            print("⚠️ PartnerSubscriptionNotificationService: Utilisateur non authentifié")
            return
        }
        
        // Arrêter l'ancien listener
        partnerListener?.remove()
        
        partnerListener = Firestore.firestore()
            .collection("users")
            .document(currentUser.uid)
            .addSnapshotListener { [weak self] snapshot, error in
                if let error = error {
                    print("❌ PartnerSubscriptionNotificationService: Erreur listener: \(error)")
                    
                    // Si erreur de permissions, gérer la déconnexion
                    if error.localizedDescription.contains("permissions") {
                        print("⚠️ PartnerSubscriptionNotificationService: Erreur de permissions - Arrêt du listener")
                        self?.partnerListener?.remove()
                        self?.partnerListener = nil
                    }
                    return
                }
                
                guard let data = snapshot?.data() else { return }
                
                // Vérifier si l'utilisateur a hérité d'un abonnement
                // 🔧 CORRECTION: Utiliser le bon nom de champ (subscriptionSharedFrom au lieu de subscriptionInheritedFrom)
                if let inheritedFrom = data["subscriptionSharedFrom"] as? String,
                   !inheritedFrom.isEmpty,  // 🔧 CORRECTION: Vérifier que le partnerId n'est pas vide
                   let isSubscribed = data["isSubscribed"] as? Bool,
                   isSubscribed {
                    Task {
                        await self?.handleSubscriptionInherited(from: inheritedFrom)
                    }
                }
                
                // Vérifier si l'utilisateur a perdu son abonnement hérité
                if let wasSubscribed = data["isSubscribed"] as? Bool,
                   !wasSubscribed,
                   let _ = data["subscriptionExpiredAt"] {
                    Task {
                        await self?.handleSubscriptionRevoked()
                    }
                }
            }
    }
    
    private func handleSubscriptionInherited(from partnerId: String) async {
        // 🔧 CORRECTION: Vérification robuste du partnerId
        guard !partnerId.isEmpty, partnerId.trimmingCharacters(in: .whitespacesAndNewlines).count > 0 else {
            print("❌ PartnerSubscriptionNotificationService: partnerId vide ou invalide: '\(partnerId)'")
            return
        }
        
        do {
            print("🔍 PartnerSubscriptionNotificationService: Récupération nom partenaire: '\(partnerId)'")
            // Récupérer le nom du partenaire qui a partagé
            let partnerDoc = try await Firestore.firestore()
                .collection("users")
                .document(partnerId)
                .getDocument()
            
            if let partnerData = partnerDoc.data(),
               let partnerName = partnerData["name"] as? String {
                
                await MainActor.run {
                    self.partnerName = partnerName
                    self.shouldShowSubscriptionInheritedMessage = true
                    print("🎁 PartnerSubscriptionNotificationService: Affichage message héritage de: \(partnerName)")
                }
            }
            
        } catch {
            print("❌ PartnerSubscriptionNotificationService: Erreur récupération nom partenaire: \(error)")
        }
    }
    
    private func handleSubscriptionRevoked() async {
        // Pour les révocations, on peut utiliser un nom générique ou récupérer depuis l'historique
        await MainActor.run {
            self.revokedPartnerName = NSLocalizedString("generic_partner", comment: "Generic partner name")
            self.shouldShowSubscriptionRevokedMessage = true
            print("🔒 PartnerSubscriptionNotificationService: Affichage message révocation")
        }
    }
    
    func dismissInheritedMessage() {
        print("🎁 PartnerSubscriptionNotificationService: Fermeture message héritage")
        shouldShowSubscriptionInheritedMessage = false
        partnerName = ""
    }
    
    func dismissRevokedMessage() {
        print("🔒 PartnerSubscriptionNotificationService: Fermeture message révocation")
        shouldShowSubscriptionRevokedMessage = false
        revokedPartnerName = ""
    }
    
    deinit {
        partnerListener?.remove()
    }
} 

File: Services/PartnerSubscriptionNotificationService.kt
package com.yourapp.services

import android.content.Context
import android.util.Log
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.yourapp.R
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

/**
 * Service de notification des changements d'abonnement partenaire.
 * iOS: ObservableObject/@Published -> Android: StateFlow pour intégration Compose.
 */
object PartnerSubscriptionNotificationService {

    private const val TAG = "PartnerSubNotifService"

    // ---- State exposé à l’UI (Compose) ----
    private val _shouldShowSubscriptionInheritedMessage = MutableStateFlow(false)
    val shouldShowSubscriptionInheritedMessage = _shouldShowSubscriptionInheritedMessage.asStateFlow()

    private val _partnerName = MutableStateFlow("")
    val partnerName = _partnerName.asStateFlow()

    private val _shouldShowSubscriptionRevokedMessage = MutableStateFlow(false)
    val shouldShowSubscriptionRevokedMessage = _shouldShowSubscriptionRevokedMessage.asStateFlow()

    private val _revokedPartnerName = MutableStateFlow("")
    val revokedPartnerName = _revokedPartnerName.asStateFlow()

    // ---- Infra ----
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var partnerListener: ListenerRegistration? = null
    private lateinit var appContext: Context

    /**
     * À appeler une fois (ex: Application.onCreate ou LaunchedEffect dans l’écran root)
     */
    fun init(context: Context) {
        appContext = context.applicationContext
        startListeningForPartnerSubscription()
    }

    /**
     * Démarre/Relance le listener Firestore sur le document de l’utilisateur courant.
     */
    fun startListeningForPartnerSubscription() {
        val currentUser = FirebaseAuth.getInstance().currentUser
        if (currentUser == null) {
            Log.w(TAG, "Utilisateur non authentifié – listener non démarré.")
            return
        }

        // Stop ancien listener le cas échéant
        partnerListener?.remove()

        partnerListener = FirebaseFirestore.getInstance()
            .collection("users")
            .document(currentUser.uid)
            .addSnapshotListener { snapshot, error ->
                if (error != null) {
                    Log.e(TAG, "Erreur listener: ${error.message}", error)
                    if (error.message?.contains("permission", ignoreCase = true) == true) {
                        Log.w(TAG, "Erreur de permissions – arrêt du listener")
                        partnerListener?.remove()
                        partnerListener = null
                    }
                    return@addSnapshotListener
                }

                val data = snapshot?.data ?: return@addSnapshotListener

                // Héritage d’abonnement (équiv. subscriptionSharedFrom + isSubscribed)
                val isSubscribed = (data["isSubscribed"] as? Boolean) ?: false
                val inheritedFrom = (data["subscriptionSharedFrom"] as? String)?.trim().orEmpty()
                if (inheritedFrom.isNotEmpty() && isSubscribed) {
                    scope.launch { handleSubscriptionInherited(inheritedFrom) }
                }

                // Révocation (isSubscribed == false && presence de subscriptionExpiredAt)
                val wasSubscribed = (data["isSubscribed"] as? Boolean) ?: false
                val hasExpiredFlag = data.containsKey("subscriptionExpiredAt")
                if (!wasSubscribed && hasExpiredFlag) {
                    scope.launch { handleSubscriptionRevoked() }
                }
            }
    }

    /**
     * Récupère le nom du partenaire qui a partagé l’abonnement et déclenche l’affichage.
     */
    private suspend fun handleSubscriptionInherited(partnerId: String) {
        val id = partnerId.trim()
        if (id.isEmpty()) {
            Log.e(TAG, "partnerId vide ou invalide: '$partnerId'")
            return
        }

        try {
            Log.d(TAG, "Récupération du nom partenaire: '$id'")
            val doc = FirebaseFirestore.getInstance()
                .collection("users")
                .document(id)
                .get()
                .await()

            val name = (doc.data?.get("name") as? String).orEmpty()
            if (name.isNotEmpty()) {
                _partnerName.value = name
                _shouldShowSubscriptionInheritedMessage.value = true
                Log.d(TAG, "Affichage message héritage de: $name")
            }
        } catch (t: Throwable) {
            Log.e(TAG, "Erreur récupération nom partenaire", t)
        }
    }

    /**
     * Définit un nom générique via strings.xml et déclenche l’affichage.
     * (Remplace NSLocalizedString(...) côté iOS)
     */
    private suspend fun handleSubscriptionRevoked() {
        val generic = if (::appContext.isInitialized) {
            appContext.getString(R.string.generic_partner)
        } else {
            "Partenaire"
        }
        _revokedPartnerName.value = generic
        _shouldShowSubscriptionRevokedMessage.value = true
        Log.d(TAG, "Affichage message révocation")
    }

    // ---- Actions UI ----
    fun dismissInheritedMessage() {
        _shouldShowSubscriptionInheritedMessage.value = false
        _partnerName.value = ""
        Log.d(TAG, "Fermeture message héritage")
    }

    fun dismissRevokedMessage() {
        _shouldShowSubscriptionRevokedMessage.value = false
        _revokedPartnerName.value = ""
        Log.d(TAG, "Fermeture message révocation")
    }

    /**
     * À appeler à la déconnexion si tu veux libérer le listener.
     */
    fun clear() {
        partnerListener?.remove()
        partnerListener = null
    }
}
